src/orion/serving/storage_resource.py

```diff
@@ -0,0 +1,232 @@
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+
+
+def _gen_host_file(basename="dump"):
+    """Generate a temporary file where data could be saved.
+
+    Create an empty file without collision in working directory.
+    Return name of generated file.
+    """
+    index = 0
+    while True:
+        file_name = f"{basename}{index}.pkl"
+        try:
+            with open(file_name, "x"):
+                return file_name
+        except FileExistsError:
+            index += 1
+            continue
+
+
+class Notifications:
+    """Stream handler to collect messages in a shared queue.
+
+    Used instead of stdout in import progress to capture log messages.
+    """
+
+    def __init__(self):
+        """Initialize with a shared queue."""
+        self.queue = multiprocessing.Queue()
+
+    def write(self, buf: str):
+        """Write received data"""
+        for line in buf.rstrip().splitlines():
+            self.queue.put(line)
+
+    def flush(self):
+        """Placeholder to flush data"""
+
+
+class ImportTask:
+    """Wrapper to represent an import task. Used to monitor task progress.
+
+    Properties:
+    - task_id: task ID, used to identify the task in web API
+    - notifications: stream handler with shared queue to capture task messages
+    - progress_message: latest progress message
+    - progress_value: latest progress (0 <= floating value <= 1)
+    - completed: shared status: 0 for running, -1 for failure, 1 for success
+
+    There is two ways to monitor task:
+    - either get messages collected in stream handler queue.
+      Stream handler collects all messages printed or logged in task.
+    - either regularly check latest message in progress_message.
+      Progress message only describe the latest step running in task.
+    """
+
+    # String representation of task status
+    IMPORT_STATUS = {0: "active", -1: "error", 1: "finished"}
+
+    def __init__(self):
+        self.task_id = str(uuid.uuid4())
+        self.notifications = Notifications()
+        self.progress_message = multiprocessing.Array("c", 512)
+        self.progress_value = multiprocessing.Value("d", 0.0)
+        self.completed = multiprocessing.Value("i", 0)
+
+    def set_progress(self, message: str, progress: float):
+        self.progress_message.value = message.encode()
+        self.progress_value.value = progress
+
+    def is_completed(self):
+        """Return True if task is completed"""
+        return self.completed.value
+
+    def set_completed(self, success=True):
+        """Set task terminated status"""
+        self.completed.value = 1 if success else -1
+
+
+def _import_data(task: ImportTask, storage, load_host, resolve, name, version):
+    """Function to run import task.
+
+    Set stream handler, launch load_database and set task status.
+    """
+    try:
+        print("Import starting.", task.task_id)
+        logging.basicConfig(stream=task.notifications, force=True, level=logging.INFO)
```

Comment:
Why do we need to set the logging level here? Is it because we lose it when running `_import_data` in a child process?

Reply:
Yes! Logging level is not shared across processes by default, because it does not use shared objects (like `multiprocessing.Queue`). So, we need to set logging level each time a process starts. In this line we also set stream as a shared object here, and then collect logging from process and access it in another process.

Reply:
It may be good to pass the logging level so it can be the same as the global one in the main process.